src/rustup.rs

```diff
@@ -0,0 +1,142 @@
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Error};
+
+use crate::config::Channel;
+use crate::{run, Context};
+
+impl Context {
+    /// Promote a `rustup` release
+    ///
+    /// The [release process] for `rustup` involves copying existing artifacts from one S3 bucket to
+    /// another, updating the manifest, and archiving the artifacts for long-term storage.
+    ///
+    /// `rustup` uses different branches to manage releases. Whenever a commit is pushed to the
+    /// `stable` branch in [rust-lang/rustup], GitHub Actions workflows build release artifacts and
+    /// copy them into `s3://dev-static-rust-lang-org/rustup/dist/`.
+    ///
+    /// When a new release is done and this method is invoked, it downloads the artifacts from that
+    /// bucket (which must always be set as the `DOWNLOAD_BUCKET` variable). A copy of the artifacts
+    /// is archived in `s3://${UPLOAD_BUCKET}/rustup/archive/${version}/`, where `version` is passed
+    /// to this program as a command-line argument. `UPLOAD_BUCKET` can either be the `dev-static`
+    /// or the `static` bucket.
+    ///
+    /// If the release is for the `stable` channel, the artifacts are also copied to the `dist/`
+    /// path in the `UPLOAD_BUCKET` bucket. The `dist/` path is used by the `rustup` installer to
+    /// download the latest release.
+    ///
+    /// Then, the `release-stable.toml` manifest is updated with the new version and copied to
+    /// `s3://${UPLOAD_BUCKET}/rustup/release-stable.toml`.
+    ///
+    /// [release process]: https://rust-lang.github.io/rustup/dev-guide/release-process.html
+    /// [rust-lang/rustup]: https://github.com/rust-lang/rustup
+    pub fn promote_rustup(&mut self) -> anyhow::Result<()> {
+        println!("Checking channel...");
+        if self.config.channel != Channel::Stable && self.config.channel != Channel::Beta {
```

Comment:
This feels really weird, since the current environments for `rustup` are `dev-static` and `prod`.

Reply:
Why are we making the channel be the condition here? Promote release already has dev static and static, as two separate environments - should be able to ignore channels I'd expect. (In a manner similar to promote branches ignoring them iirc)

Reply:
Without channels, how would you distinguish between beta and stable releases? Only be setting the respective environment variables (e.g. `UPLOAD_BUCKET`)?